tracer/src/Datadog.Trace/Debugger/DynamicInstrumentation.cs

```diff
@@ -30,182 +28,149 @@
-            ISnapshotUploader snapshotUploader,
-            IDebuggerUploader diagnosticsUploader,
-            IDebuggerUploader symbolsUploader,
-            IProbeStatusPoller probeStatusPoller,
-            ConfigurationUpdater configurationUpdater,
-            IDogStatsd dogStats)
+        internal void Initialize()
         {
-            lock (GlobalLock)
+            var originalState = Interlocked.CompareExchange(ref _initState, 1, 0);
+
+            // If we weren't in "not initialized" state, return early
+            if (originalState != 0)
             {
-                return Instance ??=
-                           new LiveDebugger(
-                               settings: settings,
-                               serviceName: serviceName,
-                               discoveryService: discoveryService,
-                               remoteConfigurationManager: remoteConfigurationManager,
-                               lineProbeResolver: lineProbeResolver,
-                               snapshotUploader: snapshotUploader,
-                               diagnosticsUploader: diagnosticsUploader,
-                               symbolsUploader: symbolsUploader,
-                               probeStatusPoller: probeStatusPoller,
-                               configurationUpdater: configurationUpdater,
-                               dogStats: dogStats);
+                return;
             }
-        }
 
-        public async Task InitializeAsync()
-        {
-            lock (GlobalLock)
+            try
             {
-                if (!CanInitialize())
+                var disabled =
+                    !(_settings.DynamicInstrumentationEnabled == true ||
+                      _settings.DynamicSettings.DynamicInstrumentationEnabled == true) ||
+                    _settings.DynamicInstrumentationEnabled == false ||
+                    _settings.DynamicSettings.DynamicInstrumentationEnabled == false;
+
+                if (disabled)
                 {
+                    Log.Warning("Dynamic Instrumentation is disabled. To enable it, please set DD_DYNAMIC_INSTRUMENTATION_ENABLED environment variable to 'true'.");
+                    // Reset to "not initialized"
+                    Interlocked.Exchange(ref _initState, 0);
                     return;
                 }
 
-                IsInitialized = true;
-            }
+                if (!Volatile.Read(ref _isRcmAvailable))
+                {
+                    Log.Warning("Dynamic Instrumentation could not be enabled because Remote Configuration Management is not available. Please ensure that you are using datadog-agent version 7.41.1 or higher, and that Remote Configuration Management is enabled in datadog-agent's yaml configuration file.");
+                    // Reset to "not initialized"
+                    Interlocked.Exchange(ref _initState, 0);
+                    return;
+                }
```

Comment:
We have a potential race here. `_isRcmAvailable` is being assigned thanks to a callback from discovery service, but we might land in this line before it had the chance to update it's value.

Reply:
Fixed in 1e7bb906196c5084b7b2d8df5e0652b22f9f6f3d